src/parser/error.rs

```diff
@@ -12,31 +15,79 @@ pub enum LexicalError {
     Generic(usize, usize),
 }
 
+/// Error indicating that a construct is not allowed when trying to interpret an `UniRecord` as a
+/// record type in a strict way.
+///
+/// See [`UniRecord::into_type_strict`](crate::parser::uniterm::UniRecord::into_type_strict).
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub enum InvalidRecordTypeError {
+    /// The record type had an invalid field, for example because it had a contract,
+    /// an assigned value, or lacked a type annotation.
+    InvalidField(RawSpan),
+    /// The record had an ellipsis.
+    IsOpen(RawSpan),
+    /// The record type had a field whose name used string interpolation.
+    InterpolatedField(RawSpan),
+    /// A field name was repeated.
+    RepeatedField { orig: RawSpan, dup: RawSpan },
+}
+
+impl InvalidRecordTypeError {
+    pub fn labels(&self) -> Vec<Label<FileId>> {
+        let label = |span: &RawSpan| {
+            Label::secondary(span.src_id, span.start.to_usize()..span.end.to_usize())
+        };
+        match self {
+            InvalidRecordTypeError::InvalidField(pos) => {
+                vec![label(pos).with_message("invalid field for a record type literal")]
+            }
+            InvalidRecordTypeError::IsOpen(pos) => {
+                vec![label(pos).with_message("cannot have ellipsis in a record type literal")]
+            }
+            InvalidRecordTypeError::InterpolatedField(pos) => {
+                vec![label(pos).with_message("this field uses interpolation")]
+            }
+            InvalidRecordTypeError::RepeatedField { orig, dup } => {
+                vec![
+                    label(orig).with_message("first occurrence"),
+                    label(dup).with_message("second occurrence"),
+                ]
+            }
+        }
+    }
+
+    pub fn notes(&self) -> Option<String> {
+        match self {
+            InvalidRecordTypeError::InvalidField(_) => Some(
+                "Value assignments such as `<field> = <expr>`, and metadata \
+                    annotation (annotation, documentation, etc.) are forbidden."
+                    .into(),
+            ),
+            InvalidRecordTypeError::InterpolatedField(_) => {
+                Some("String interpolation in field names is forbidden in record types".into())
```

Comment:
As a side note, this one is a bit subtle to classify. In RFC002, we say something like "if something is a record type (in the old syntax), then it is parsed as a static type, otherwise it's a record literal". Given that interpolation is forbidden inside record types, taking this literally would mean that we shouldn't complain that "a record type has interpolation", but rather see this record as a record literal. However, we fail in this case too, because we forbid (since 1.0) type annotation on fields without definition.

So the semantics (as far as failing is considered) is the same, but not the error message. In the end I wonder if we shouldn't mention both cause, because we can't really know if the user wanted to write a record type with interpolation or a record literal with a field with a static type annotation but without a definition. Maybe adding a note corresponding to the latter could be sufficient? Something in the spirit of "and a field with a type annotation and without a definition is forbidden in record literals".

Reply:
I've updated the error message to explain why we interpreted the literal as a type (one of two reasons: presence of a tail, or all fields have type annotations only). This is also relevant for the duplicated field name error.